File: java-sdk-example/src/test/java/CreateSandboxTestWithCustomPatch.java
import com.signadot.ApiClient;
import com.signadot.ApiException;
import com.signadot.api.SandboxesApi;
import com.signadot.model.*;
import io.restassured.RestAssured;
import io.restassured.builder.RequestSpecBuilder;
import io.restassured.http.ContentType;
import io.restassured.specification.RequestSpecification;
import org.apache.commons.lang3.RandomStringUtils;
import org.testng.annotations.AfterSuite;
import org.testng.annotations.BeforeSuite;
import org.testng.annotations.Test;

import java.util.Arrays;
import java.util.List;

import static io.restassured.RestAssured.given;
import static org.hamcrest.Matchers.*;

public class CreateSandboxTestWithCustomPatch {

  public static final String HOTROD = "hotrod";
  public static final String CUSTOM_PATCH = "spec:\n" +
                                            "  template:\n" +
                                            "    spec:\n" +
                                            "      containers:\n" +
                                            "      - name: hotrod\n" +
                                            "        env:\n" +
                                            "        - name:  PATCH_TEST_VAR\n" +
                                            "          value: foo\n";

  public static final String ORG_NAME = System.getenv("SIGNADOT_ORG");
  public static final String SIGNADOT_API_KEY = System.getenv("SIGNADOT_API_KEY");
  public static final String CLUSTER_NAME = System.getenv("SIGNADOT_CLUSTER_NAME");

  private static RequestSpecification requestSpec;

  ApiClient apiClient;
  SandboxesApi sandboxesApi;
  CreateSandboxResponse response;
  String sandboxID;

  @BeforeSuite
  public void createSandboxWithCustomPatch() throws ApiException, InterruptedException {
    apiClient = new ApiClient();
    apiClient.setApiKey(SIGNADOT_API_KEY);
    sandboxesApi = new SandboxesApi(apiClient);

    String sandboxName = String.format("custom-patch-test-%s", RandomStringUtils.randomAlphanumeric(5));

    SandboxFork frontendFork = new SandboxFork()
      .forkOf(new ForkOf().kind("Deployment").namespace(HOTROD).name("frontend"))
      .customizations(new SandboxCustomizations()
        .addImagesItem(new Image().image("signadot/hotrod:49aa0813feba0fb74e4edccdde27702605de07e0"))
      ).addEndpointsItem(new ForkEndpoint().name("hotrod-frontend").port(8080).protocol("http"));

    SandboxFork routeFork = new SandboxFork()
      .forkOf(new ForkOf().kind("Deployment").namespace(HOTROD).name("customer"))
      .customizations(new SandboxCustomizations()
        .addImagesItem(new Image().image("signadot/hotrod:49aa0813feba0fb74e4edccdde27702605de07e0"))
        .addEnvItem(new EnvOp().name("FROM_TEST_VAR").valueFrom(
          new EnvValueFrom().fork(
            new EnvValueFromFork().forkOf(
              new ForkOf()
                .kind("Deployment")
                .namespace("hotrod")
                .name("frontend")
            ).expression("{{ .Service.Host }}:{{ .Service.Port }}")
          )
        ))
        .patch(new CustomPatch().type("strategic").value(CUSTOM_PATCH)))
      .addEndpointsItem(new ForkEndpoint().name("hotrod-customer").port(8081).protocol("http"));

    CreateSandboxRequest request = new CreateSandboxRequest()
      .cluster(CLUSTER_NAME)
      .name(sandboxName)
      .description("Java SDK: sandbox creation with custom patch example")
      .addForksItem(routeFork)
      .addForksItem(frontendFork);

    response = sandboxesApi.createNewSandbox(ORG_NAME, request);

    sandboxID = response.getSandboxID();
    if (sandboxID == null || sandboxID.isBlank()) {
      throw new RuntimeException("Sandbox ID not set in API response");
    }

    List<PreviewEndpoint> endpoints = response.getPreviewEndpoints();
    if (endpoints.size() == 0) {
      throw new RuntimeException("preview endpoints not available in API response");
    }

    PreviewEndpoint endpoint = null;
    for (PreviewEndpoint ep : endpoints) {
      if ("hotrod-customer".equals(ep.getName())) {
        endpoint = ep;
        break;
      }
    }
    if (endpoint == null) {
      throw new RuntimeException("No endpoint found for route service");
    }

    // set the base URL for tests
    RestAssured.baseURI = endpoint.getPreviewURL();

    RequestSpecBuilder builder = new RequestSpecBuilder();
    builder.setBaseUri(endpoint.getPreviewURL());
    builder.addHeader("signadot-api-key", SIGNADOT_API_KEY);

    requestSpec = builder.build();

    // Check for sandbox readiness
    while (!sandboxesApi.getSandboxReady(ORG_NAME, sandboxID).isReady()) {
      Thread.sleep(5000);
    }
  }

  @Test
  public void testCustomEnvVars() {
    given()
      .spec(requestSpec)
      .when()
      .get("/customer?customer=392")
      .then()
      .statusCode(200)
      .assertThat()
      .body("PatchVar", equalTo("foo"))
      .body("FromVar", not(is(emptyOrNullString())));
  }

  @AfterSuite
  public void deleteSandbox() throws ApiException {
    sandboxesApi.deleteSandboxById(ORG_NAME, sandboxID);
  }
}

File: python-sdk-example/tests/integration/create_sandbox_test_with_custom_patch.py
from __future__ import print_function

import os
import random
import string
import time
import unittest

import requests
import timeout_decorator

from signadot_sdk import Configuration, SandboxesApi, ApiClient, SandboxFork, SandboxForkOf, \
    SandboxCustomizations, SandboxImage, SandboxForkEndpoint, Sandbox, SandboxSpec, SandboxCustomPatch, SandboxTTL
from signadot_sdk.rest import ApiException


def get_random_string(length):
    return ''.join(random.choice(string.ascii_lowercase + string.digits) for i in range(length))


class TestBasic(unittest.TestCase):
    CUSTOM_PATCH = """
    spec:
      template:
        spec:
          containers:
          - name: hotrod
            env:
            - name: PATCH_TEST_VAR
              value: "{}"
    """
    HOTROD_TEST_IMAGE = "signadot/hotrod:49aa0813feba0fb74e4edccdde27702605de07e0"
    SIGNADOT_ORG = os.getenv("SIGNADOT_ORG")
    CLUSTER_NAME = os.getenv("SIGNADOT_CLUSTER_NAME")
    SIGNADOT_API_KEY = os.getenv('SIGNADOT_API_KEY')
    if SIGNADOT_ORG is None:
        raise OSError("SIGNADOT_ORG is not set")
    if CLUSTER_NAME is None:
        raise OSError("SIGNADOT_CLUSTER_NAME is not set")
    if SIGNADOT_API_KEY is None:
        raise OSError("SIGNADOT_API_KEY is not set")

    configuration = Configuration()
    configuration.api_key['signadot-api-key'] = SIGNADOT_API_KEY

    sandboxes_api = SandboxesApi(ApiClient(configuration))
    sandbox_name = None
    env_var_value = None
    endpoint_url = None
    headers_dict = {"signadot-api-key": SIGNADOT_API_KEY}

    @classmethod
    @timeout_decorator.timeout(120)
    def setUpClass(cls):
        cls.env_var_value = get_random_string(5)

        customer_fork = SandboxFork(
            fork_of=SandboxForkOf(
                kind="Deployment",
                name="customer",
                namespace="hotrod"
            ),
            customizations=SandboxCustomizations(
                images=[
                    SandboxImage(image=cls.HOTROD_TEST_IMAGE)
                ],
                patch=SandboxCustomPatch(
                    type="strategic",
                    value=cls.CUSTOM_PATCH.format(cls.env_var_value)
                )
            ),
            endpoints=[
                SandboxForkEndpoint(
                    name="hotrod-customer",
                    port=8081,
                    protocol="http"
                )
            ]
        )
        cls.sandbox_name = "custom-patch-test-{}".format(get_random_string(5))
        request = Sandbox(
            spec=SandboxSpec(
                description="Python SDK: sandbox creation with custom patch example",
                ttl=SandboxTTL(duration="10m"),
                cluster=cls.CLUSTER_NAME,
                forks=[customer_fork]
            )
        )

        try:
            sandbox = cls.sandboxes_api.apply_sandbox(cls.SIGNADOT_ORG, cls.sandbox_name, request)
        except ApiException as e:
            print("Exception creating a sandbox: %s\n" % e)

        filtered_endpoints = list(filter(lambda ep: ep.name == "hotrod-customer", sandbox.endpoints))
        if len(filtered_endpoints) == 0:
            raise RuntimeError("Endpoint `hotrod-customer` missing")
        cls.endpoint_url = filtered_endpoints[0].url

        # Code block to wait until sandbox is ready
        while not sandbox.status.ready:
            time.sleep(5)
            sandbox = cls.sandboxes_api.get_sandbox(cls.SIGNADOT_ORG, cls.sandbox_name)

    def test_valid_result(self):
        service_url = "{}/customer?customer=392".format(self.endpoint_url)
        response = requests.get(service_url, headers=self.headers_dict)
        response.raise_for_status()
        data = response.json()

        self.assertEqual(data["PatchVar"], self.env_var_value, "PatchVar must equal {}".format(self.env_var_value))

    @classmethod
    def tearDownClass(cls):
        cls.sandboxes_api.delete_sandbox(cls.SIGNADOT_ORG, cls.sandbox_name)


if __name__ == '__main__':
    unittest.main()
